### PR TITLE
[IMP] Website_slide: use built_in methods for placeholder images

### PR DESCRIPTION
**Pull request by angv-odoo**
Before this commit placeholder images were generated in multiple
different ways in website_slide. This commit makes use of
_get_placeholder_filename function and standardizes the way
these placeholder images are generated for website_slide.

Task-2908029

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by flch-odoo — COMMENTED** [review]
Preliminary comments ;-)

File: addons/website_slides/static/src/scss/website_slides_profile.scss [hunk 1]
```diff
@@ -2,8 +2,4 @@
 .o_wprofile_slides_course_card_body {
     cursor: pointer;
 
-    .o_wslides_background_image {
-        background-size: cover;
-        background-position: center;
-    }
```
Comment (flch-odoo) [inline addons/website_slides/static/src/scss/website_slides_profile.scss:8]: Is this removing a duplication? or the actual rules (because we do still use `o_wslides_background_image`)

File: addons/website_slides/static/src/scss/website_slides.scss [hunk 2]
```diff
@@ -514,11 +514,9 @@ $line-height-truncate: 1.25em;
 // **************************************************
 .o_wslides_lesson_main {
     .o_wslides_lesson_aside {
-        .o_wslides_background_image_aside_card {
```
Comment (flch-odoo) [inline addons/website_slides/static/src/scss/website_slides.scss:517]: Is it not a "background" image? I wouldn't rename it unless it's confusing at some level.

File: addons/website_slides/models/slide_channel.py [hunk 3]
```diff
@@ -468,14 +467,11 @@ def _compute_partner_has_new_content(self):
...
-        for channel in self:
-            channel.website_background_image_url = (
-                channel.website_id.image_url(channel, 'image_256') if channel.image_1920
-                else '/website_slides/static/src/img/channel-%s-default.jpg'
-                     % ('training' if channel.channel_type == 'training' else 'documentation')
-            )
+    def _get_placeholder_filename(self, field):
+        image_fields = ['image_%s' % size for size in [1920, 1024, 512, 256, 128]]
+        if field in image_fields:
+            return 'website_slides/static/src/img/channel-%s-default.jpg' % ('training' if self.channel_type == 'training' else 'documentation')
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:473]: We could keep `website_background_image_url` and use it here
```suggestion
            return self.website_background_image_url
```
This way, it is immediately usable in `slide.slide`'s `_get_placeholder_filename` as it is now defined.
Otherwise (if we remove the field, and I agree it *is* cleaner, then you need an upgrade script as well), we'd have to port that logic in that `slide.slide` version to get the channel image as first fallback if defined (and not the placeholder of the channel image).

File: addons/website_slides/views/website_slides_templates_lesson.xml [hunk 4]
```diff
@@ -99,9 +99,8 @@
...
 <template id="slide_aside_card" name="Related Slide">
     <a class="list-group-item list-group-item-action d-flex align-items-start px-2" t-att-href="'/slides/slide/%s' % (slug(aside_slide))">
-        <t t-set="slide_image" t-value="website.image_url(aside_slide, 'image_256') if can_access_channel or aside_slide.is_preview else slide.channel_id.website_background_image_url"/>
-        <div class="me-1 border o_wslides_background_image_aside_card" t-attf-style="background-image: url(#{slide_image});"/>
-        <div class="overflow-hidden d-flex flex-column justify-content-start">
+            <div t-field="aside_slide.image_1920" t-options="{'widget': 'image', 'style': 'width: 64px;height: 64px;object-fit:cover', 'preview_image': 'image_128'}" class="me-1 border"/>
+            <div class="overflow-hidden d-flex flex-column justify-content-start">
             <h6 t-esc="aside_slide.name" class="o_wslides_desc_truncate_2 mb-1" style="line-height: 1.15"/>
             <small class="text-600">
                 <t t-esc="aside_slide.total_views"/> Views &#8226; <timeago class="timeago" t-att-datetime="aside_slide.create_date"></timeago>
```
Comment (flch-odoo) [inline addons/website_slides/views/website_slides_templates_lesson.xml:106]: Could be unindented one level.

File: addons/website_slides/models/slide_slide.py [hunk 5]
```diff
@@ -250,6 +250,9 @@ def _compute_is_new_slide(self):
         for slide in self:
             slide.is_new_slide = slide.date_published > fields.Datetime.now() - relativedelta(days=7) if slide.is_published else False
 
+    def _get_placeholder_filename(self, field):
+        return self.channel_id._get_placeholder_filename(field)
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_slide.py:254]: Sorry if last review's comment wasn't clear enough. See above for details.

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 6]
```diff
@@ -122,7 +122,9 @@
                             <!-- ==== Header Left ==== -->
                             <div class="col-12 col-md-4 col-lg-3 position-relative">
                                 <div class="d-flex align-items-end justify-content-around h-100">
-                                    <div t-field="channel.image_1920" t-options='{"widget": "image", "class": "o_wslides_course_pict d-inline-block mb-2 mt-3 my-md-0"}' class="h-100"/>
+                                    <div t-field="channel.image_1920" t-options="{'widget': 'image', 'class' : 'o_wslides_course_pict d-inline-block mb-2 mt-3 my-md-0', 'preview_image': 'image_1024'}" class=" h-100">
```
Comment (flch-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:125]:
```suggestion
                                    <div t-field="channel.image_1920" t-options="{'widget': 'image', 'class' : 'o_wslides_course_pict d-inline-block mb-2 mt-3 my-md-0', 'preview_image': 'image_1024'}" class="h-100">
```

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 7]
```diff
@@ -122,7 +122,9 @@
...
+                                        <t t-if="channel.partner_has_new_content" t-call="website_slides.course_card_information"/>
```
Comment (flch-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:126]: Unless discussed with @LNAodoo , I wouldn't add this here either (not pretty imho).

**Review by flch-odoo — COMMENTED** [review]
Hey @angv-odoo ,
Here is a second round of comments. 
Please also fix the second commit up into the first one as they have no distinct role.
Also, very importantly, I see no image for non previewable slides of `/slides/trees-wood-and-gardens-3` as public user...

File: addons/website_slides/models/slide_channel.py [hunk 8]
```diff
@@ -477,6 +471,12 @@ def _compute_website_background_image_url(self):
                      % ('training' if channel.channel_type == 'training' else 'documentation')
             )
 
+    def _get_placeholder_filename(self, field):
+        image_fields = ['image_%s' % size for size in [1920, 1024, 512, 256, 128]]
+        if field in image_fields:
+            return self._compute_website_background_image_url()
+        return super()._get_placeholder_filename(field)
+
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:479]: If we move the method (which makes sense to fit guidelines), it shouldn't go here (between compute methods) but below, in `MISC`. (maybe even in last position?).

File: addons/website_slides/models/slide_channel.py [hunk 9]
```diff
@@ -477,6 +471,12 @@ def _compute_website_background_image_url(self):
                      % ('training' if channel.channel_type == 'training' else 'documentation')
             )
 
+    def _get_placeholder_filename(self, field):
+        image_fields = ['image_%s' % size for size in [1920, 1024, 512, 256, 128]]
+        if field in image_fields:
+            return self._compute_website_background_image_url()
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:477]: You don't need to call the method (which doens't return anything), but the field directly.

File: addons/website_slides/models/slide_channel.py [hunk 10]
```diff
@@ -477,6 +471,12 @@ def _compute_website_background_image_url(self):
                      % ('training' if channel.channel_type == 'training' else 'documentation')
             )
 
+    def _get_placeholder_filename(self, field):
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:474]: With this rewrite,
* You can now call it in slide.slide as was previously present
* _compute_website_background_image_url needs to be updated (and renamed?), because we'll only be calling it to get the default image (when we know there is no image_1920.

**Review by flch-odoo — COMMENTED** [review]
@angv-odoo , please let me know if I misunderstood something or if some of my comments are unclear :-)

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 11]
```diff
@@ -747,14 +740,10 @@
...
         <a t-if="can_access" t-attf-href="/slides/slide/#{slug(slide)}#{query_string}" t-title="slide.name" style="height:150px">
-            <div t-if="slide.image_1920" t-field="slide.image_1920" t-options="{'widget': 'image', 'preview_image': 'image_512'}"
-                 class="o_wslides_background_image h-100"/>
-            <div t-else="" class="o_wslides_background_image">
-                <img t-att-src="channel.website_background_image_url" class="img img-fluid"/>
-            </div>
+            <div t-field="slide.image_1920" t-options="{'widget': 'image', 'preview_image': 'image_512'}" class="o_wslides_background_image h-100"/>
         </a>
         <div t-else="" class="o_wslides_background_image" style="height:150px">
-            <img t-att-src="channel.website_background_image_url" class="img img-fluid"/>
```
Comment (flch-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:757]: I think we should keep the getting the channel image as placeholder for the slide if the user can't access the slide.

**Review by flch-odoo — COMMENTED** [review]
With this comment's resolution, the upgrade script, and merge of commits, it should be good to go for me :-)

File: addons/website_slides/views/website_slides_templates_course.xml (hunk 11 above)
Comment (angv-odoo) [reply addons/website_slides/views/website_slides_templates_course.xml:757]: I added the line back.

Comment (flch-odoo) [reply addons/website_slides/views/website_slides_templates_course.xml:757]: Nope, that does not work anymore. Please compare your runbot (https://21513559-master-all.runbot119.odoo.com/slides/trees-wood-and-gardens-3) and master (https://21512998-master-all.runbot144.odoo.com/slides/trees-wood-and-gardens-3) **as public user**

Comment (flch-odoo) [reply addons/website_slides/views/website_slides_templates_course.xml:757]: @angv-odoo : Comment above is **not** resolved ?

**Review by flch-odoo — COMMENTED** [review]
Hey @angv-odoo , we're getting there! 
A few more suggestions for the code, and for the commit, could you merge everything in one commit and summarize its message content, please?
After that (and rebase), I also think we should be good to go :) Stay strong!

File: addons/website_slides/models/slide_channel.py [hunk 12]
```diff
@@ -477,14 +471,10 @@ def _compute_partner_has_new_content(self):
...
-    def _compute_website_background_image_url(self):
+    @api.depends('channel_type')
+    def _compute_website_default_background_image_url(self):
         for channel in self:
-            channel.website_background_image_url = (
-                channel.website_id.image_url(channel, 'image_256') if channel.image_1920
-                else '/website_slides/static/src/img/channel-%s-default.jpg'
-                     % ('training' if channel.channel_type == 'training' else 'documentation')
-            )
+            channel.website_default_background_image_url = 'website_slides/static/src/img/channel-{self.channel_type}-default.jpg'
```
Comment (flch-odoo) [inline addons/website_slides/models/slide_channel.py:477]: :-)
```suggestion
            channel.website_default_background_image_url = f'website_slides/static/src/img/channel-{self.channel_type}-default.jpg'
```

File: addons/website_slides/static/src/scss/website_slides.scss [hunk 13]
```diff
@@ -232,8 +232,8 @@ $line-height-truncate: 1.25em;
 // **************************************************
 .o_wslides_home_main {
     .o_wslides_home_aside_loggedin {
-        background: none;
         @include media-breakpoint-up(lg) {
+            background: none;
```
Comment (flch-odoo) [inline addons/website_slides/static/src/scss/website_slides.scss:236]: Is this change necessary or a remnant of something else?

File: addons/website_slides/static/src/scss/website_slides.scss [hunk 14]
```diff
@@ -530,7 +533,11 @@ $line-height-truncate: 1.25em;
                 top: -58px;
             }
         }
-
+        .o_wslides_lesson_side_card{
```
Comment (flch-odoo) [inline addons/website_slides/static/src/scss/website_slides.scss:536]: I think we're missing blank lines before and after this block, to match the rest of the file?

**Review by flch-odoo — COMMENTED** [review]
Hello @angv-odoo , It ***is*** looking good to me. 
Only two minor comments left, plus one about the commit message: do we still have something related to the test in here?
Cheers!

**Comment by tde-banana-odoo** [comment]
@robodoo r+
